File: android/src/main/java/com/nuvopoint/flutter_nfc_acs/DeviceScanner.java
package com.nuvopoint.flutter_nfc_acs;

import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;

import io.flutter.plugin.common.EventChannel.EventSink;
import io.flutter.plugin.common.EventChannel.StreamHandler;

import static android.content.ContentValues.TAG;
import static com.nuvopoint.flutter_nfc_acs.FlutterNfcAcsPlugin.ERROR_NO_PERMISSIONS;

class DeviceScanner extends BluetoothPermissions implements StreamHandler {
  private final HashMap<String, String> btDevices = new HashMap<>();
  private final BluetoothAdapter bluetoothAdapter;
  private EventSink events;
  private final Handler handler;
  private boolean scanning = false;
  private final Activity activity;

  private static final long SCAN_PERIOD = 10000;

  DeviceScanner(@NonNull BluetoothAdapter adapter, @NonNull Activity activity) {
    bluetoothAdapter = adapter;
    this.activity = activity;
    handler = new Handler();
  }

  @Override
  public void onListen(Object arguments, EventSink events) {
    this.events = events;
    // TODO: Ask for permissions to turn on bluetooth if disabled.

    if (!hasPermissions()) {
      requestPermissions();
      return;
    }

    if (!btDevices.isEmpty()){
      events.success(btDevices);
    }

    startScan();
  }

  @Override
  public void onCancel(Object arguments) {
    stopScan();
    events = null;
  }

  /* Device scan callback. */
  private BluetoothAdapter.LeScanCallback mLeScanCallback = (device, rssi, scanRecord) -> {
    if (events != null) {
      new Handler(Looper.getMainLooper()).post(() -> {
        if (!btDevices.containsKey(device.getAddress())) {
          btDevices.put(device.getAddress(), device.getName());
          events.success(btDevices);
        }
      });
    } else {
      Log.w(TAG, "Could not output devices, because the event sink was null");
    }
  };

  private void startScan() {
    handler.postDelayed(() -> {
      if (scanning) {
        scanning = false;
        bluetoothAdapter.stopLeScan(mLeScanCallback);
      }
    }, SCAN_PERIOD);

    scanning = true;
    bluetoothAdapter.startLeScan(mLeScanCallback);
  }

  private void stopScan() {
    bluetoothAdapter.stopLeScan(mLeScanCallback);
    scanning = false;
  }

  @Override
  protected Activity getActivity() {
    return activity;
  }

  @Override
  protected void afterPermissionsGranted() {
    if (bluetoothAdapter != null) {
      startScan();
    } else {
      Log.e(TAG, "Bluetooth adapter was null, in the permission callback");
    }
  }

  @Override
  protected void afterPermissionsDenied() {
    if (events != null) {
      events.error(ERROR_NO_PERMISSIONS, "Location permissions are required", null);
      events = null;
    }
  }
}

File: android/src/main/kotlin/com/tillty/flutter_nfc_acs/DeviceScanner.kt
package com.tillty.flutter_nfc_acs

import android.annotation.SuppressLint
import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.ContentValues
import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import java.util.*

internal class DeviceScanner(private val adapter: BluetoothAdapter, override val activity: Activity) :
  BluetoothPermissions(), EventChannel.StreamHandler {
  private var btDevices: HashMap<String, String>? = null
  private var events: EventSink? = null
  private var scanning = false
  private val scanner: BluetoothLeScanner?
    get() = adapter.bluetoothLeScanner

  companion object {
    private const val SCAN_PERIOD: Long = 10000
  }

  override fun onListen(arguments: Any?, events: EventSink?) {
    this.events = events
    // TODO: Ask for permissions to turn on bluetooth if disabled.
    if (hasPermissions()) {
      requestPermissions()
      return
    }
    startScan()
  }

  override fun onCancel(arguments: Any?) {
    stopScan()
    events = null
  }

  private val scanCallback = object : ScanCallback() {
    @SuppressLint("MissingPermission")
    override fun onScanResult(callbackType: Int, result: ScanResult) {
      val e = events
      if (e != null) {
        Handler(Looper.getMainLooper()).post {
          val device = result.device
          if (!btDevices!!.containsKey(device.address)) {
            btDevices!![device.address] = device.name ?: "N/A"
            e.success(btDevices)
          }
        }
      } else {
        Log.w(
          ContentValues.TAG,
          "Could not output devices, because the event sink was null"
        )
      }
    }
  }

  @SuppressLint("MissingPermission")
  private fun startScan() {
    btDevices = HashMap()
    scanning = true

    Handler(Looper.getMainLooper()).postDelayed({
      if (scanning) {
        stopScan()
      }
    }, SCAN_PERIOD)
    scanner?.startScan(scanCallback)
  }

  @SuppressLint("MissingPermission")
  private fun stopScan() {
    scanner?.stopScan(scanCallback)
    scanning = false
  }

  override fun afterPermissionsGranted() {
    startScan()
  }

  override fun afterPermissionsDenied() {
    if (events != null) {
      events!!.error(
        FlutterNfcAcsPlugin.ERROR_NO_PERMISSIONS,
        "Location permissions are required",
        null
      )
      events = null
    }
  }
}